src/process-snippet.js

```diff
@@ -1,117 +1,82 @@
-    map.addMapping({
-      generated: {
-        line: decoratedCodeFirstLine + i,
-        column: 0,
-      },
-      original: {
-        line: position.start.line + i + 1,
-        column: 0,
-      },
-      source,
-    })
-  }
-
-  map.setSourceContent(source, code)
+const { SourceMapConsumer, SourceMapGenerator } = require('source-map')
 
-  return map.toString()
-}
+const injectDependencies = require('./transformers/inject-dependencies')
+const importToRequire = require('./transformers/import-to-require')
 
 const sourceMapsPrefix =
   '//# sourceMappingURL=data:application/json;charset=utf-8;base64,'
 
-// injectDependencies :: Object -> Object -> String -> String
-const injectDependencies = (file, code, position, deps, globals) => {
-  const depsStr = listDependencies(deps || [])
-  const globalsStr = listGlobals(globals || {})
-  const sourceMaps = generateSourceMaps(
-    file,
-    code,
-    position,
-    depsStr,
-    globalsStr
+const mergeMaps = (m1, m2) => {
+  if (!m1) {
+    return m2
+  }
+
+  const map = SourceMapGenerator.fromSourceMap(new SourceMapConsumer(m2))
+  map.applySourceMap(new SourceMapConsumer(m1))
+  return map.toJSON()
+}
+
+/**
+ * Take a string with code and a list of transformers and return the
+ * transformed code and the composed sourcemaps
+ */
+const applyTransforms = (filename, inputCode, transformers) => {
+  const { code, map } = transformers.reduce(
+    (previousResult, transformer) => {
+      if (previousResult.code === null) {
+        return previousResult
+      }
+
+      const result = transformer(filename, previousResult.code)
+
+      if (result.code === null) {
+        return result
+      }
+
+      return {
+        code: result.code,
+        map: mergeMaps(previousResult.map, result.map),
+      }
+    },
+    {
+      code: inputCode,
+      map: null,
+    }
   )
-  const sourceMapsBase64 = Buffer.from(sourceMaps).toString('base64')
 
+  return { code, map }
+}
+
+const addInlineSourcemap = (code, map) => {
+  const sourceMapsBase64 = Buffer.from(JSON.stringify(map)).toString('base64')
   const sourceMapsInline = `${sourceMapsPrefix}${sourceMapsBase64}`
 
-  return [
-    "require('source-map-support').install()",
-    '// snippet dependencies',
-    depsStr,
-    globalsStr,
-    '',
-    code,
-    sourceMapsInline,
-  ].join('\n')
+  const transformedCode = `${code}\n${sourceMapsInline}`
+
+  return transformedCode
 }
 
-const processSnippet = (file, code, position, deps, globals) =>
-  injectDependencies(file, code, position, deps, globals)
+const processSnippet = (file, code, position, deps, globals) => {
+  const injectDependenciesTransformer = injectDependencies(
+    position,
+    deps,
+    globals
+  )
+  const importToRequireTransformer = importToRequire()
+
+  const result = applyTransforms(file, code, [
+    injectDependenciesTransformer,
+    importToRequireTransformer,
+  ])
+
+  if (result.code === null) {
+    // Do something with result.error
```

Comment:
The transformer contract is like this:

```(filename, code) -> {code, map, error}```

If the transformer was able to parse it, it will return `{code, map}`. If not, it will return `{code: null, error`.

If some transformer failed, it makes no sense to continue with the pipeline. The final result will have a null code and the first error.

The best thing here would be to indicate where the syntax error is, with the proper line in the readme, but I'm not sure how to do that.

Reply:
This is really nice. What I would change is the `code === null`, just check if there's an error instead.

About handling the error. The `snippet` object the runner receives should have that error property and the runner can check that before running.

This is what the runner passes to the reporter

https://github.com/Huemul/chimi/blob/5482ef05f4ebbef1001f4562748cd2e6ef372b33/src/run-snippet.js#L48-L56

We can still resolve that promise but never run the snippet, just pass the error and other stuff.